wicketd/src/mgs_manager.rs

```diff
@@ -0,0 +1,155 @@
+/// MgsManager and not result in a new underlying request to MGS. This keeps
+/// interaction to a constant amount of work, and limits bursts to any updates
+/// that require further pulling from MGS.
+///
+/// Update interacions from Wicket (e.g. HTTP POST/PUT requests to wicketd)
+/// will be forwarded one at a time to MGS. We only allow one outstanding
+/// update at a time. If the update request is to actually update software,
+/// and there is a long running update in progress, the request will not
+/// be forwarded and a relevant response will be returned to wicket.
+pub struct MgsManager {
+    log: Logger,
+    tx: mpsc::Sender<MgsRequest>,
+    rx: mpsc::Receiver<MgsRequest>,
+    mgs_client: gateway_client::Client,
+    // Remove the need to copy a potentially sizeable amount of data
+    inventory: Arc<RackV1Inventory>,
+}
+
+impl MgsManager {
+    pub fn new(log: &Logger, mgs_addr: SocketAddrV6) -> MgsManager {
+        let log = log.new(o!("component" => "wicketd MgsManager"));
+        let (tx, rx) = mpsc::channel(CHANNEL_CAPACITY);
+        let endpoint =
+            format!("http://[{}]:{}", mgs_addr.ip(), mgs_addr.port());
+        info!(log, "MGS Endpoint: {}", endpoint);
+        let mgs_client = gateway_client::Client::new(&endpoint, log.clone());
+        let inventory = Arc::new(RackV1Inventory::default());
+        MgsManager { log, tx, rx, mgs_client, inventory }
+    }
+
+    pub fn get_handle(&self) -> MgsHandle {
+        MgsHandle { tx: self.tx.clone() }
+    }
+
+    pub async fn run(mut self) {
+        let mgs_client = self.mgs_client;
+        let mut inventory_rx = poll_inventory(&self.log, mgs_client).await;
+
+        loop {
+            tokio::select! {
+                // Poll MGS inventory
+                Some(sps) = inventory_rx.recv() => {
+                    self.inventory = Arc::new(RackV1Inventory {
+                        sps: sps.into_iter().map(|sp| (sp, vec![])).collect()
+                    });
+                }
+
+                // Handle requests from clients
+                Some(request) = self.rx.recv() => {
+                    debug!(self.log, "{:?}", request);
+                     match request {
+                         MgsRequest::GetInventory {reply_tx, ..} => {
+                            let _ = reply_tx.send(self.inventory.clone());
+                         }
+                     }
+                }
+            }
+        }
+    }
+}
+
+pub async fn poll_inventory(
+    log: &Logger,
+    client: gateway_client::Client,
+) -> mpsc::Receiver<Vec<SpInfo>> {
+    let log = log.clone();
+
+    // We only want one outstanding inventory request at a time
+    let (tx, rx) = mpsc::channel(1);
+
+    tokio::spawn(async move {
+        let mut ticker = interval(MGS_POLL_INTERVAL);
+        ticker.set_missed_tick_behavior(MissedTickBehavior::Delay);
+        loop {
+            ticker.tick().await;
+            match client.sp_list(MAX_COMPONENTS, None, Some(MGS_TIMEOUT)).await
```

Comment:
I know we already chatted about this a bit and are going to chat more, but just noting as I'm reading - this only fetches the first page of data.

Reply:
@jgallagher and I discussed this, and there's an open issue for converting this MGS endpoint to an unpaginated one. I'm going to do that next. https://github.com/oxidecomputer/omicron/issues/1913